### ISSUE TITLE
Undefined, null, NaN etc. should not be considered as "normal" values

### ISSUE THREAD

**Issue by kavi87** [messageformat/messageformat#209]
BUG:

Currently if a variable is `null` or `undefined`  or `NaN` (which we'll call `nil`), messageformat considers that it is the value to stringify in the result. 

The way to currently handle the issue is to use a `select` and handle the `nil case. That's terribly cumbersome in my opinion. 

It should be possible to configure a default replacement for `nil`.

An even better way in my opinion would be to allow syntax that would completely remove part of the expression if the value is not available in the context:

`{ firstName } { lastName } [[ - work at { company } ]] // If company is nil, remove stuff between [[ ]]`

This would require relaxing the conformity with ICU standard.

**Comment by eemeli** [comment]
Have I understood right that in summary you're asking that instead of having to write this:
```
{firstName} {lastName} {select, company, undefined{} other{- work at {company}}}
```
you would like us to break conformity with the ICU spec by enabling this equivalent syntax?
```
{firstName} {lastName} [[- work at {company}]]
```

Sorry, but what you're asking for is far too much for far too little gain. If the `select` option isn't working for you, I'd recommend you write a [custom formatter](https://messageformat.github.io/messageformat/page-guide#toc11__anchor) that would enable you to use something like this:
```
{firstName} {lastName} {notEmpty, company, - work at {company}}
```

**Comment by kavi87** [comment]
How should someone write { firstName } alone without having it print "undefined" ? { firstName, select, undefined{... ? 

What I should point out is that in my situation the values are entered by users. I think { firstName } is what you should write and you should handle missing context information because it is obvious that from a business point of view, the value of firstName is NOT undefined, that's just JS informing you that your value is not defined. So you should be able to catch it and replace it.

**Comment by kavi87** [comment]
Currently I'm using your parser to get the token of type arguments and check their presence against the context.

 The parser is required anyway for us since the translations are dynamic and must come from user defined value at runtime.

As for my approach to remove useless part, I do not think it breaks ICU in any way. It just extends it. That would require a richer parser that's all.

Also the lib is very good and thanks for your work.

**Comment by eemeli** [comment]
Can you clarify what you mean by "values are entered by users"? Do users input messageformat source directly? Maybe you could add a wrapper around the compiler that would replace empty values with your expected string, such as `"{firstName}"`. For that, you'll likely need to use the [AST parser](/messageformat/parser).

Globally changing how undefined or null values are handled would break quite a bit of existing code, so something like that would require really good reasons for implementation.

The ICU MessageFormat spec defines which characters are special, so "extending" it by making `[[]]` special would indeed break it.

For handling `NaN` values and other edge cases, I would again recommend that you write a custom formatter as I mentioned earlier; then you can implement exactly the logic you need.

**Comment by kavi87** [comment]
Yes, the users enters their translation values. For instance for the i18n key "app.home.welcome", the value could be "Hi { subject_name }". We provide the authenticated user its context for the keys (that's why I would also have found interesting for them to be able to do stuff like { subject.account.type }, instead I've resorted to flatten the objects to have it as subject_account_type).

I've circumvented the event with the parser alone, it gives you the arguments token so I can check if the wanted property is actually defined on the context and add a default value if not. I can work around to strip the unwanted remaining piece of strings by allowing an escaped character as a delimiter (my [[).

In the end you are probably right, ICU is a standard and your library should conform for most people. Hopefully that discussion can help someone with the same issues and ways to work around.

Thanks !
